New linear solver factory

loumalouomega opened the pull request:
I need help to migrate and improve the new_linear_solver_factory

loumalouomega committed:
Improve new_linear_solver_factory.py

loumalouomega commented:
I started with the StructuralMechanicsApplication, but there are many applications that can be migrated too. Right now the current `linear_solver_factory.py` has more options than the `new_linear_solver_factory.py`. It will be nice if people that know how this solvers work can help me with the migration

RiccardoRossi commented:
Hi @loumalouomega ,
it is not exactly as you say. The point is that the "old" one has all the options in a singe file (the factory itself).

the new one implements the parsing of the parameters in the constructor of each linear solver.

at the very end the options should be the same

RiccardoRossi commented:
still let's follow @ipouplana's comment and let's ma

RiccardoRossi commented:
sorry i wanted to delete my comment and closed the issue instead. 

i was telling "let's follow @ipouplana's comment and mention everyone (@KratosMultiphysics) in this issue since it may affect a lot of people

pooyan-dadvand commented:
I'm ok with this PR. Nevertheless if we want to make a new factory I would suggest to use the factory pattern instead of a monolithic if based file we have now.

RiccardoRossi commented:
Hi @pooyan-dadvand, in principle i would also agree in the long run to implement a better method for this. The definition of what is a factory pattern sees however to depend a lot on the language and on wether the language implements or not "reflection" (which in this is context is to be intended as the possibility of querying all the subclasses of a given class).

reading into your mind (please correct me if i am wrong) my understanding is that you propose to implement the Factory in c++ rather than in python.

If this is correct, than your proposal would correspond to 
1 -  add a base class LinearSolverFactory with a static singleton "linear_solver_map<name, prototype_obj>" and register all the linear solvers into linear_solver_map.
2 - add to each of the linear solvers a function "MakeSolver" which returns a created object. (also we would need the empty constructor for all the solvers)
3 - define a method MakeSolver of the base class which acts equivalently to the currently existing one.

To my understanding the biggest advantages would be:
 
1. possibility to call the function from within c++ 
2. availability (in c++, in python this could be done already) to get the list of all available solvers.

rubenzorrilla reviewed (commented):
I'm ok with the enhancements. However, I would take profit of this PR to rename the "new_linear_solver_factory" to "linear_solver_factory" and the old one to "deprecated_linear_solver_factory" to avoid this kind of misunderstandings in the future.

pooyan-dadvand commented:
As we discussed it just now, the idea is to have a solver factory base class:

```c++
class LinearSolverFactoryBase {
public:
    virtual LinearSolver::Pointer Create(Parameters) = 0;
}
```
and a template factory class which create a solver given as template argument:

```c++
template <typename TLinearSolverType>
class LinearSolverFactory {
public:
    LinearSolver::Pointer Create(Parameters) {
        if(Parameters["Scaling"].GetBool() == true)
            return PointerType(new ScalingSolver(new TLinearSolverType(Parameters)));
        else
            return PointerType(new TLinearSolverType(Parameters));
}
```
and use the factory for registration:
```c++
KRATOS_REGISTER_LINEAR_SOLVER("CGSolver", LinearFactorySolver<CGSolver<Space,...>>)
```

RiccardoRossi commented:
Just to say that i am convinced that making the registration in python is a good way to go. **I would propose keeping on hold** this pull request until i take a deep look in making it in c++

pooyan-dadvand commented:
@RiccardoRossi Meanwhile this new factory is an step forward to the old one and keeping this one won't harm anyone until we make it more organic.

loumalouomega commented:
I will revert this commit in my branch then

RiccardoRossi commented:
wait, don't do anything yet.

just keep it on hold

loumalouomega commented:
No, I meant in my branch, that one that is taller and with more commits than the Babylon tower

RiccardoRossi committed:
registration of linear solvers in c++

RiccardoRossi committed:
adding the registration mechanism to the base class of linear solver

RiccardoRossi committed:
adding registration mechanism for preconditioners

RiccardoRossi committed:
calling functions to register linear solvers and preconditioners

RiccardoRossi committed:
adding interfaces to the construction of preconditioners given the parameters and correcting a use of Parameters by passing a reference instead of a copy of the parameters object

RiccardoRossi committed:
cleaning up. atually this file is no longer needed and should be removed asap

RiccardoRossi commented:
with the last commit I did the registration of the linear solvers in c++ as @pooyan-dadvand  suggested

this is **NOT a urgent commit** and should be definitely be postponed to after having finalized the release process (that is, let's leave this for september).

essentially the object linear_solver_factory does not need any longer to be included in python, since it is substituted by 

      solver = KratosMultiphysics.LinearSolverFactoryBase().CreateSolver(configuration)

loumalouomega committed:
CammelCase and nullptr in pastix_comples_solver.h

In C++11 the C NULL can be replaced by nullptr. Besides adding descriptions to the methods

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega committed:
CammelCase

RiccardoRossi commented on kratos/linear_solvers/linear_solver.h:
```diff
@@ -234,8 +234,8 @@ class LinearSolver
         VectorType& rB,
-        typename ModelPart::DofsArrayType& rdof_set,
-        ModelPart& r_model_part
+        typename ModelPart::DofsArrayType& rDoFSet,
```
this is defined in all the solvers. I would not change this

loumalouomega committed:
Removing unnecesary KRATOS_ERROR in base class

loumalouomega commented:
What do we do with this PR?

loumalouomega committed:
Merging

RiccardoRossi commented:
I think that we need some discussion on this. Guys, can you take a look and give feedback?

pooyan-dadvand commented on kratos/includes/linear_solver_factory.h:
```diff
@@ -0,0 +1,89 @@
+
+
+
+void RegisterPreconditioners();
```
Why these are global functions?

RiccardoRossi replied:
where would you put them?

pooyan-dadvand replied:
Each registration in the corresponding header/cpp file

pooyan-dadvand commented on kratos/includes/linear_solver_factory.h:
```diff
@@ -0,0 +1,89 @@
+
+
+void RegisterPreconditioners();
+void RegisterLinearSolvers();
```
And this one?

RiccardoRossi replied:
same, let's agree of where to put them

pooyan-dadvand commented on applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp:
```diff
@@ -119,7 +119,33 @@ void  AddLinearSolversToPython()
+        
+    KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("GMRESSolver"), GMRESSolverFactory);
+    KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("SuperLUSolver"), SuperLUSolverFactory);
+    KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("SuperLUIterativeSolver"), SuperLUIterativeSolverFactory);
```
We may find a better place to register the linear solvers. i.e. a new `register_linear_solvers.cpp` file

pooyan-dadvand commented on kratos/linear_solvers/linear_solver.h:
```diff
@@ -406,7 +407,50 @@ class LinearSolver
+// Class LinearSolver///here we add the functions needed for the registration
+template< typename TSparseSpace, typename TlocalSpace>
+class LinearSolverFactoryBase
+{
```
It would be better to move this class to a separate file.

RiccardoRossi replied:
the point is that this way the class is available wherever you include the linear solver, which was what i wanted to achieve

pooyan-dadvand commented on kratos/linear_solvers/linear_solver.h:
```diff
@@ -406,7 +407,50 @@ class LinearSolver
+            KRATOS_ERROR << "Trying to construct a Linear solver with solver_type= " << Settings["solver_type"].GetString() << std::endl << "which does not exist. The list of available options (for currently loaded applications) is: " << std::endl <<
+                         KratosComponents< LinearSolverFactoryBase<TSparseSpace,TlocalSpace> >() << std::endl;
+        }
+        const auto& aux = KratosComponents< LinearSolverFactoryBase<TSparseSpace,TlocalSpace> >::Get( Settings["solver_type"].GetString()  );
```
`factory` would be a better name for variable `aux`

pooyan-dadvand commented on kratos/python_scripts/new_linear_solver_factory.py:
```diff
@@ -1,82 +1,8 @@
-
 def ConstructSolver(configuration):
     import KratosMultiphysics
+    return KratosMultiphysics.LinearSolverFactoryBase().CreateSolver(configuration)
```
Now I'm confused, I think what referred as base is the factory itself and what is referred as factory is a concrete factory.

pooyan-dadvand commented on kratos/python_scripts/new_linear_solver_factory.py:
```diff
@@ -1,82 +1,8 @@
-        linear_solver = KratosMultiphysics.AMGCLSolver(configuration)
-    elif(solver_type == "AMGCL_NS_Solver"):
-        linear_solver = KratosMultiphysics.AMGCL_NS_Solver(configuration)
-        
```
Lot of ifs removed!! 😄

pooyan-dadvand commented on kratos/sources/kernel.cpp:
```diff
@@ -36,6 +38,9 @@ namespace Kratos
         mKratosApplication.RegisterVariables();
+        
+        RegisterLinearSolvers();
+        RegisterPreconditioners();
```
I would strongly avoid calling these functions in the constructor of the Kernel. They are not exception free and with a lot of background details.....

RiccardoRossi replied:
I see your point. I think it was needed to put it there, but i don't remember the details. i'll check when i have some time

pooyan-dadvand reviewed (changes requested):
In general I liked very much the way @RiccardoRossi implemented the new linear solver factory and now is much more extensible than before. However I would suggest some minor changes to improve it:
- We should find some standard places for registering the linear solvers and preconditioners
- I got confused by the LinearSolverFactoryBase as it seems to be the factory to be used and not the base. 
- I would strongly avoid registering in constructor of Kernel

RiccardoRossi commented on kratos/linear_solvers/preconditioner.h:
```diff
@@ -403,6 +402,63 @@ inline std::ostream& operator << (std::ostream& OStream,
 ///@}
 
 
+template< typename TSparseSpace, typename TlocalSpace>
```
same argument as for the linear_solver_base. I am putting it in the same archive to have it available where the preconditioenr is available

pooyan-dadvand commented:
I agree that we need to give it the final iteration after release.

pooyan-dadvand commented:
Just to add that the overall design is a significant improvment

RiccardoRossi committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

RiccardoRossi committed:
moving call to register to the variables.cpp instead of keeping it in kernel.cpp

RiccardoRossi committed:
removing use of Parameters::Pointer

RiccardoRossi committed:
accepting changes

RiccardoRossi committed:
attempting to define a register linear solvers function. Some problems with multiply defined symbols in c-libraries solved by defining the class body in the same file as add_linear_solvers_to_python. Not optimal but i could not find a better solution

RiccardoRossi commented:
@msandre @philbucher @oberbichler here i registered the real-valued solvers but not the complex-valued once, since i was not sure of what should be defined. Could u take a look? it should be trivial to also register the complex valued ones once you know which ones should be registered. 

also note that onces this is done, you could construct the complex linear solver you want directly from c++ within the class FEAST solver simply passing the pertinent parameters

pooyan-dadvand commented:
@RiccardoRossi I think we should create a new `Registry` class to centralize all registrations in Kratos. Actually this message is reminder note for myself! :smile:

oberbichler commented:
Who is using the complex solvers?

Is it possible to register the solvers from within the application and without modifying the …factory.py?

RiccardoRossi commented:
@pooyan-dadvand let's discuss your proposal on monday. Also to be honest i don't understand how you would like the current registration mechanism to be changed, so i did not address some of your suggestions

@oberbichler yes, the idea is to register the solvers at the c++ level, so that the _factory.py becomes trivial. The main advantages are that registration is done at compile time and that the list of available solvers is automatically updated. Also it is useful to have the possibility of using the factory from c++

RiccardoRossi committed:
adding missing file

philbucher commented:
@qaumann is using complex solvers afaik

qaumann commented:
@philbucher The FEAST solver needs a complex linear solver (see fest_solver.h). I am not using a complex solver outside of FEAST

RiccardoRossi committed:
removing a warning

msandre commented:
There are two complex solvers for FEAST:

- A default one which uses the skyline from amgcl. 

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_linear_solvers_to_python.cpp#L212

- a faster one that depends on the tpl pastix:

https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp#L102

we need to add both to the registry. It should simply be a matter of registering the solvers and running the eigen tests in structural mechanics application to see that it works. I can't do it today but will try to do it soon.

Commit:
small formatting.

Commit:
added ublas space aliases.

Commit:
make scaling solver compile for complex scalars.

Commit:
fix warning.

Commit:
added complex solver factory.

Commit:
fixed solver names in structural mechanics tests.

msandre commented:
The complex solvers are added now. @philbucher the check you added to Eigen_Q4_Thick_2x2_Plate_test_parameters.json fails on my machine. It looks like the numbers are affected by roundoff. Can you look at it?

msandre commented:
Also @RiccardoRossi can you check my changes to the scaling solver? I had to modify it so it would compile with the complex solvers.

philbucher commented:
@msandre I know that it fails, fixing it is on my todo list

RiccardoRossi commented:
ping @pooyan-dadvand

RiccardoRossi commented on applications/ExternalSolversApplication/external_includes/feast_solver.h:
```diff
@@ -101,9 +101,9 @@ class FEASTSolver: public LinearSolver<TSparseSpaceType, TDenseSpaceType,
+        mParam.RecursivelyValidateAndAssignDefaults(default_params);
 
-        if (mpParam->GetValue("linear_solver_settings")["solver_type"].GetString() != "complex_skyline_lu_solver")
+        if (mParam.GetValue("linear_solver_settings")["solver_type"].GetString() != "skyline_lu")
```
i review myself since a lot of time passed: should this be complex_skyline_lu_solver?

RiccardoRossi commented on kratos/sources/linear_solver_factory.cpp:
```diff
@@ -0,0 +1,123 @@
+        KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("AMGCLSolver"), AMGCLSolverFactory); //registered with two different names
+        KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("AMGCL_NS_Solver"),AMGCL_NS_SolverFactory );
+        KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("ScalingSolver"),ScalingSolverFactory );
+        KratosComponents<ComplexLinearSolverFactoryBaseType>::Add(std::string("SkylineLUComplexSolver"), SkylineLUComplexSolverFactory);
```
@pooyan-dadvand ,maybe we should call this 

       SkylineLUComplexSolverFunctor

instead of 

       SkylineLUComplexSolverFactory

RiccardoRossi commented on kratos/sources/variables.cpp:
```diff
@@ -586,7 +588,8 @@ namespace Kratos
   void KratosApplication::RegisterVariables()
   {
-
+        RegisterLinearSolvers();
```
i am not sure if this is the correct place, however we need somewher that is surely called at initialization time

RiccardoRossi reviewed (commented):
@pooyan-dadvand i read the book you sent me and i would say that the implementation is ok.

most probably we could rename some stuff to "Functor" and call the FactoryBase as "AbstractFactory" but other than that i think that the implementation is more or less ok.

pooyan-dadvand commented:
let me check it again, I forgot my comments on this

loumalouomega commented:
Which is the status of this?

pooyan-dadvand commented:
I should give it the last polish with @RiccardoRossi but after the release

RiccardoRossi commented:
@pooyan-dadvand this is getting stale. It would be nice if this could be in for release 6...

pooyan-dadvand commented:
I agree with you. The delay is my fault. Let see it together next week and finish it.

roigcarlo commented:
Hi, this is flagged for 6.0 but I see a lot of conflicts here. I will remove this from the release. If this is ready for 6.0 please re add it next week and we will make a separate Pr into the fixed branch

pooyan-dadvand commented on applications/ExternalSolversApplication/CMakeLists.txt:
```diff
@@ -47,6 +47,7 @@ endif(${INCLUDE_FEAST} MATCHES ON)
 ## generate variables with the sources
 set( KRATOS_EXTERNAL_SOLVERS_APPLICATION_SOURCES    
 	${CMAKE_CURRENT_SOURCE_DIR}/externalsolvers_application.cpp 
+	${CMAKE_CURRENT_SOURCE_DIR}/custom_utilities/register_linear_solvers.cpp 
```
I would put file in the linear_solvers folder. Meanwhile it is almost empty...

pooyan-dadvand commented on applications/ExternalSolversApplication/custom_python/add_linear_solvers_to_python.cpp:
```diff
@@ -113,10 +112,59 @@ void  AddLinearSolversToPython()
+
+
+//Must put this definition here to avoid a problem with multiply defined symbols when including the external C libraries
+RegisterLinearSolvers::RegisterLinearSolvers()
```
At this stage you don't need the `RegisterLinearSolvers `at all. You may do it in application registration until I implement the Registry class in the core.

pooyan-dadvand commented on applications/ExternalSolversApplication/custom_utilities/register_linear_solvers.h:
```diff
@@ -0,0 +1,207 @@
+//    |  /           |
```
You don't need this class as long as using `KratosComponents `as your registry.

pooyan-dadvand commented on applications/ExternalSolversApplication/externalsolvers_application.cpp:
```diff
@@ -27,6 +28,8 @@ void KratosExternalSolversApplication::Register()
     KratosApplication::Register();
     std::cout << "Initializing KratosExternalSolversApplication... " << std::endl;
+    
+    RegisterLinearSolvers();
```
At this stage you may put your the class constructor directly here.

pooyan-dadvand commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -422,7 +422,7 @@ def _create_convergence_criterion(self):
 
     def _create_linear_solver(self):
-        import linear_solver_factory
+        import new_linear_solver_factory as linear_solver_factory
```
It is not a substitution of the old one? without new prefix I mean.

RiccardoRossi replied:
actually here i would call directly

       KratosMultiphysics.LinearSolverFactoryBase().CreateSolver(configuration)

pooyan-dadvand commented on kratos/linear_solvers/bicgstab_solver.h:
```diff
@@ -85,9 +85,16 @@ class BICGSTABSolver : public IterativeSolver<TSparseSpaceType, TDenseSpaceType,
+        BaseType(settings) 
+    {
+        if(settings.Has("preconditioner_type"))
+            BaseType::SetPreconditioner( PreconditionerFactoryBase<TSparseSpaceType,TDenseSpaceType>().CreatePreconditioner(settings["preconditioner_type"].GetString()) );
```
I would make the `CreatePreconditioner `a method of the `BaseType` (even with the check `Has("preconditioner_type")` included.

pooyan-dadvand replied:
And here I would call the CreatePreconditioner in the base class constructor to avoid code duplication.

pooyan-dadvand reviewed (changes requested):
Now I understood the structure and I think that the confusing part (for me) was the fact that in your implementation the ObjectFactory is the LinearSolverFactoryBase and the ObjectCreator is the LinearSolverFactory derived from the base one. Meanwhile to have the work reusable after the implementation of Registery and ObjectFactory, it would be better to have Create(Parameters) methods for each solver. Nevertheless I understand if you decide to postponed this change to when we have the registery.

They are some minor comments also in the code.

loumalouomega commented:
I created this PR, but I think that most of that changes are from @RiccardoRossi , should I do something?

pooyan-dadvand commented:
My comments are for @RiccardoRossi

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega commented:
I solved the conflicts to tackle this again, I hope I haven't broke anything.

loumalouomega commented:
Some leftovers of boost python are needed to be cleaned

loumalouomega committed:
Boost dependency

loumalouomega committed:
Duplicated input

loumalouomega committed:
Typo

loumalouomega committed:
Removing boost dependency

loumalouomega committed:
Missing retrocompatibility names

loumalouomega committed:
Adding EigenSolvers to factory

loumalouomega committed:
Retrocompatibility name

loumalouomega committed:
More retrocompatibility name

loumalouomega commented:
@RiccardoRossi and @pooyan-dadvand I fixed the compilation and now all the tests work. I added the EigenApplicationSolvers too

loumalouomega committed:
Clean (I)

loumalouomega committed:
Clean (II)

loumalouomega commented:
Ok, right now the only remaining issue is the cotire compilation, @RiccardoRossi can you check that?

philbucher commented:
you sure its cotire?
You are missing some overrides which are turned into errors

loumalouomega commented:
that explains a lot :), I was reading cotire messages, sorry for the noise

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega committed:
Missing override

philbucher commented:
when Travis doesn't compile take a close look at the output => since #2572  some warnings are turned into errors in order to keep the code clean

loumalouomega commented:
OK

loumalouomega commented:
This is ready for merge

oberbichler commented:
👍 Looks good!

About the complex solvers in the `EigenSolverApplication`: I think they are testest (@armingeiser ?)

Maybe we should drop the old solver names.

oberbichler commented:
...and why `std::string("...")` instead of `"..."`?

loumalouomega commented:
Where is that std::string?

oberbichler commented on applications/EigenSolversApplication/custom_python/add_custom_solvers_to_python.cpp:
```diff
@@ -29,69 +32,128 @@ namespace Python
+    static auto SparseQRFactory= LinearSolverFactory<SpaceType,LocalSpaceType,SparseQRType>();
+    static auto ComplexSparseQRFactory= LinearSolverFactory<ComplexSpaceType,ComplexLocalSpaceType,ComplexSparseQRType>();
+
+    KratosComponents<LinearSolverFactoryBaseType>::Add(std::string("SparseLUSolver"), SparseLUFactory);
```
here

armingeiser commented:
@oberbichler The complex solvers from the `EigenSolversApplication` are only tested within FEAST (the same way as also ComplexSkylineLU).

oberbichler commented:
👍  I was just wondering why `ComplexLDLT/LLT` are commented out but they are not covered by the FEAST test...

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega committed:
Removing std::string

loumalouomega committed:
Minor

loumalouomega commented:
Done

RiccardoRossi reviewed (commented):
@loumalouomega to me this looks ok. Thank you very much for taking the time to fix this.

@pooyan-dadvand it is on you to approve this ...

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega committed:
@RiccardoRossi comment

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega commented:
I solved the conflict with AMatrix, maybe we should merge this before more conflicts arrise

loumalouomega committed:
Clean

loumalouomega committed:
Reverting Ublas definition

TODO: Complex number to be compatibilized with AMatrix @pooyan-dadvand

loumalouomega commented:
@pooyan-dadvand this is ready

loumalouomega commented:
Do I apply the changes related with AMatrix to this branch or do I create a new PR after this is merged?

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega commented:
Conflict solved @pooyan-dadvand

loumalouomega commented:
@pooyan-dadvand I would like to have this merged in order to develop a proper adaptive NR. I discussed with @RiccardoRossi and the idea is to extend the idea of the factories to all the components integrating the NR (B&S, convergence criterion, etc...). That way we could create a constructor for the NR based on parameters instead of adding more arguments as doing right now.

pooyan-dadvand commented:
@loumalouomega we can check this together in Wednesday when I pass by CIMNE

loumalouomega commented:
Ok :+1:

pooyan-dadvand commented on kratos/includes/linear_solver_factory.h:
```diff
@@ -0,0 +1,89 @@
+
+template <typename TSparseSpace, typename TLocalSpace, typename TLinearSolverType>
+class LinearSolverFactory : public LinearSolverFactoryBase<TSparseSpace,TLocalSpace>
+{
```
`LinearSolverFactory` to be changed to `StandardLinearSolverFactory`

pooyan-dadvand commented on kratos/includes/linear_solver_factory.h:
```diff
@@ -0,0 +1,89 @@
+protected:
+
+    typename LinearSolver<TSparseSpace,TLocalSpace>::Pointer CreateHelper(Kratos::Parameters settings) const override
+    {
```
`CreateHelper` -> `CreateSolver`

pooyan-dadvand commented on kratos/linear_solvers/linear_solver.h:
```diff
@@ -416,7 +417,50 @@ class LinearSolver
+
+// Class LinearSolver///here we add the functions needed for the registration
+template< typename TSparseSpace, typename TLocalSpace>
+class LinearSolverFactoryBase
```
`LinearSolverFactoryBase` to `LinearSolverFactory`

pooyan-dadvand commented on kratos/linear_solvers/linear_solver.h:
```diff
@@ -416,7 +417,50 @@ class LinearSolver
+        return KratosComponents< LinearSolverFactoryBase<TSparseSpace,TLocalSpace> >::Has( SolverType );
+    }
+
+    virtual typename LinearSolver<TSparseSpace,TLocalSpace>::Pointer CreateSolver(Kratos::Parameters Settings)
```
`CreateSolver` to `Create`

pooyan-dadvand commented on kratos/linear_solvers/linear_solver.h:
```diff
@@ -416,7 +417,50 @@ class LinearSolver
+        return aux.CreateHelper( Settings );
+    }
+protected:
+    virtual typename LinearSolver<TSparseSpace,TLocalSpace>::Pointer CreateHelper(Kratos::Parameters Settings)  const
```
`CreateHelper` to `CreateSolver`

pooyan-dadvand commented on kratos/linear_solvers/preconditioner.h:
```diff
@@ -403,6 +402,63 @@ inline std::ostream& operator << (std::ostream& OStream,
 
 
+template< typename TSparseSpace, typename TlocalSpace>
+class PreconditionerFactoryBase
```
Same for this base

pooyan-dadvand commented on kratos/linear_solvers/scaling_solver.h:
```diff
@@ -89,34 +91,43 @@ class ScalingSolver : public LinearSolver<TSparseSpaceType, TDenseSpaceType,  TR
     
-    ScalingSolver(Parameters settings,
-                  typename LinearSolver<TSparseSpaceType, TDenseSpaceType, TReordererType>::Pointer p_linear_solver
+//     ScalingSolver(Parameters settings,
```
take out the comments please

pooyan-dadvand reviewed (changes requested):
I would add macro for registering the solvers and preconditioners and distribute them in their corresponding files

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega committed:
Change name (I)

loumalouomega committed:
Rename (II)

loumalouomega committed:
Rename (III)

loumalouomega committed:
Docu

loumalouomega committed:
Clean up

loumalouomega committed:
Adding register

loumalouomega committed:
Preconditioner factory separated

loumalouomega commented:
Some changes are still pendent. I will upload them later

loumalouomega committed:
Clean up

loumalouomega committed:
CreateSolver -> Create

loumalouomega committed:
CreateHelper -> CreateSolver

loumalouomega committed:
Separating python register factory

loumalouomega commented:
The only thing I am not able yet is to register in each separated file without conflict python register

loumalouomega committed:
Merge branch 'master' into feature-migration-new-linear-solver-factory

loumalouomega commented:
@pooyan-dadvand @roigcarlo told me that is not possible to directly register the solvers in the header, it is necessary at least to call a cpp

pooyan-dadvand reviewed (approved):
So we postpone the distributed registration to when Kratos Factory is implemented

loumalouomega commented:
Hurray

philbucher commented on kratos/solving_strategies/strategies/solving_strategy.h:
```diff
@@ -348,7 +348,6 @@ class SolvingStrategy
      */
     virtual double GetResidualNorm()
     {
-        KRATOS_ERROR << "You are calling to the base class method GetResidualNorm, please define in you derived class the method" << std::endl;
```
why did you remove this?

philbucher reviewed (commented):
Finally :D

You have not updated / registered te trilinos solvers right?

loumalouomega commented:
I think the Trilinos are not registered, I have not register them at least

pooyan-dadvand commented:
Yes, this one become even older than me! :stuck_out_tongue_winking_eye:

philbucher commented:
one question guys:
to use it for the mpi-solvers, one should derive a `DistributedLinearSolverFactory` from the `LinearSolverFactory` right?
same as what is done for the `StandardLinearSolverFactory` (weird name btw)

loumalouomega commented:
Yes, exactly, yesterday I tried to compile Trilinos and fix it, but I was unable. Are you suing Ubuntu?

pooyan-dadvand commented:
> same as what is done for the StandardLinearSolverFactory (weird name btw)

Any suggestion?

loumalouomega commented:
I will mention #3185 as this PR moves the factories to one folder, and can be an opportunity to change the name of the standard factory

loumalouomega commented:
BTW linear solver factories are added to Trilinos in here:

#3204
